### DISCUSSION TITLE
Why do slots fall back to the previously active state and not the default?

### DISCUSSION THREAD

**Discussion by amir-s** [vercel/next.js#59418] [App Router]
I'm using the app router for a dashboard i'm creating. The `Layout` has the `{props.children}`, and also a `{props.subMenu}` for a parallel route, so the sub pages can have customized sub menus on the header which is on the main layout.

<img width="770" alt="Screenshot 2023-12-08 at 2 08 18 PM" src="https://github.com/vercel/next.js/assets/905963/05e7730b-be3c-4c16-865f-24abda440479">

I thought I can just setup a `default.tsx` with `() => null` component so not all the sub pages need to define a sub menu, and if they don't, it'll be empty. But that's not how it works. The sub menu is persisted between client side navigations if I go from a page that has sub menu to a page that doesn't. Only if I do a hard reload the sub menu disappears. I thought this might be a bug but in the [docs](https://nextjs.org/docs/app/building-your-application/routing/parallel-routes) about parallel routes it says:

> 
> [Navigation](https://nextjs.org/docs/app/building-your-application/routing/parallel-routes#navigation)
> On navigation, Next.js will render the slot's previously active state, even if it doesn't match the current URL.
> 
> [Reload](https://nextjs.org/docs/app/building-your-application/routing/parallel-routes#reload)
> On reload, Next.js will first try to render the unmatched slot's default.js file. If that's not available, a 404 gets rendered.

so this is by design, but I don't understand what's the use-case for that and why it has been designed that way.

**Comment by finnmerlett** [comment]
+1000 this seems like really unintuitive behaviour. I just do not get why you should ever get different results when navigating vs visiting by url / hard reloading. Even the url interception feature seems difficult to grasp and unclear what the outcome of using it will be, but at least you have to active decide to enable that. The fact that for the inconsistent slot fallback behaviour, the docs just say *it is thusly*, and give no explanation for why or any way to turn it off _infuriates_ me.

**Comment by doomietrue** [comment]
I'm puzzled by a particular challenge: I aim to implement a Layout with slots for Header and Sidebar. The complication arises on one unique page where the Header and Sidebar have a slightly altered appearance. I wish to maintain some client-side state for the Sidebar, like the state of a collapsed accordion which in this case works. However, the issue I'm encountering is when the Sidebar is rendered with the accordion and the user moves to a different page, the transition doesn't seem to recognize the change in the route. 

The documentation is saying that "_On navigation, Next.js will render the slot's previously active state, even if it doesn't match the current URL._" but it doesn't explain why. Would be nice to have a possibility to somehow opt out of this behavior for individual slots.

**Comment by nphmuller** [comment]
I also am a bit stumped by this behavior. As a workaround I removed the default.ts fallback and implemented the slot for every route. The routes that don’t need to slot simply render a component that returns null. A bit more work, but it’s doable I guess.

**Comment by vmukhteev** [comment]
As a workaround I use `[...]/page.tsx` instead of `default.tsx`.

**Reply by stnmonroe** [reply]
I tried this one, and after an intercepted route it works going backward, but broke the appropriate slot updates moving forward. Also, was using `(...)` not `[...]`. I assume you were as well, or maybe this is why it wasn't working. I did not get any response using square brackets.

**Reply by artem-prudnikov** [reply]
@stnmonroe
I used [...] and it now works back and forth. Had to restart NextJS app for the changes to apply.

**Comment by icopp** [comment]
I agree 100%. This is intensely unintuitive behavior and is hard to design around. I feel like everyone can, or should, agree that the page rendering behavior on navigation, and on a new refresh, should always be exactly the same except for the very specific case of a feature like intercepting routes that exists entirely to preclude normal navigation.

**Comment by angrybacon** [comment]
Unless I'm missing something, the catch-all workaround requires a name

`app/@toc/[...any]/page.ts`

```tsx
export default () => null;
```

While it works at first glance, it appears that rendering the node conditionally isn't that straightforward. My use case is a table of content that should only render in routes that are defined under `@toc/`.

```tsx
export default ({ children, toc }: PropsWithChildren<{ toc: ReactNode }>) => (
  <html lang="en">
    <Box
      component="body"
      sx={[
        { ... },
        toc !== undefined && { 'footer, main': { mr: `${TOC_WIDTH}px` } },
      ]}
    >
      <AppRouterCacheProvider>
        <Header />
        <Box component="main">{children}</Box>
        {toc && <Box component="aside" sx={{ ... }}>{toc}</Box>}
        <Footer />
      </AppRouterCacheProvider>
    </Box>
  </html>
);
```

The issue there is that `toc` is obviously never undefined. However I'd rather configure styles for the `aside` element directly in `layout.tsx`. It does look like I'm not using parallel routes for their intended behavior but I could be missing something obvious?

**Reply by icopp** [reply]
> Unless I'm missing something, the catch-all workaround requires a name

It does now, but that requirement was [only recently added](https://github.com/vercel/next.js/issues/87671) and broke existing uses of `[...]`.

**Comment by hungnm-kaopiz** [comment]
Damn, I ran into an issue and had to rewrite the whole component using a different solution.
